pineappl_capi/src/lib.rs

```diff
@@ -1357,3 +1355,253 @@ pub unsafe extern "C" fn pineappl_string_delete(string: *mut c_char) {
+        .map(|s| Order {
+            alphas: s[0].try_into().unwrap(),
+            alpha: s[1].try_into().unwrap(),
+            logxir: s[2].try_into().unwrap(),
+            logxif: s[3].try_into().unwrap(),
+            logxia: s[4].try_into().unwrap(),
+        })
+        .collect();
+
+    // Bin limits
+    let bin_limits = unsafe { slice::from_raw_parts(bin_limits, bins + 1).to_vec() };
+
+    // Types of convolutions
+    let mut convolutions = Vec::new();
+    for i in 0..nb_convolutions {
+        let c_conv = unsafe {
+            let str_ptr = *convolution_types.add(i);
+            CStr::from_ptr(str_ptr)
+                .to_str()
+                .expect("Failed to convert one of the `convolution_types` into a proper string.")
+        };
+
+        let pid_value = unsafe { *pdg_ids.add(i) };
+
+        let cvtype = match c_conv {
+            "UnpolPDF" => Conv::new(ConvType::UnpolPDF, pid_value),
+            "PolPDF" => Conv::new(ConvType::PolPDF, pid_value),
+            "UnpolFF" => Conv::new(ConvType::UnpolFF, pid_value),
+            "PolFF" => Conv::new(ConvType::PolFF, pid_value),
+            _ => todo!(),
+        };
+        convolutions.push(cvtype);
+    }
+
+    // Grid interpolations
+    let mut interp_vecs = Vec::new();
+    let interp_slices = unsafe { std::slice::from_raw_parts(interpolations, nb_convolutions + 1) };
+    for interp_ref in interp_slices {
+        interp_vecs.push(construct_interpolation(interp_ref));
+    }
+
+    // Kinematics
+    let mut kinematics_vec = Vec::new();
+    for k in 0..=nb_convolutions {
+        let kin = unsafe {
+            let kin_ptr = *kinematics.add(k);
+            CStr::from_ptr(kin_ptr)
+                .to_str()
+                .expect("Failed to convert one of the `kinematics` types into a proper string.")
+        };
+
+        let kin_type = match kin {
+            "scale" => Kinematics::Scale(0),
+            "x1" => Kinematics::X1,
+            "x2" => Kinematics::X2,
+            _ => todo!(),
+        };
+        kinematics_vec.push(kin_type);
+    }
+
+    Box::new(Grid::new(
+        pid_basis,
+        channels.0.clone(),
+        orders,
+        bin_limits,
+        convolutions,
+        interp_vecs,
+        kinematics_vec,
+        Scales {
+            ren: ScaleFuncForm::Scale(0),
+            fac: ScaleFuncForm::Scale(0),
+            frg: ScaleFuncForm::NoScale,
+        },
+    ))
+}
+
+/// Fill `grid` for the given momentum fractions {`x1`, `x2`,..., `xn`}, at the scale `q2`
+/// for the given value of the `order`, `observable`, and `lumi` with `weight`.
+///
+/// # Safety
+///
+/// If `grid` does not point to a valid `Grid` object, for example when `grid` is the null pointer,
+/// this function is not safe to call.
+#[no_mangle]
+pub unsafe extern "C" fn pineappl_grid_fill2(
+    grid: *mut Grid,
+    order: usize,
+    observable: f64,
+    lumi: usize,
+    ntuple: *const f64,
+    weight: f64,
+    size_tuple: usize,
```

Comment:
Omit `size_tuple` and instead we must assume that this is given by `grid.kinematics().len()`.